Framework/Parallel/inc/MantidParallel/ThreadingBackend.h

```diff
@@ -0,0 +1,134 @@
+#include <istream>
+#include <map>
+#include <mutex>
+#include <ostream>
+#include <sstream>
+#include <tuple>
+
+namespace Mantid {
+namespace Parallel {
+namespace detail {
+
+/** ThreadingBackend provides a backend for data exchange between
+  Communicators in the case of non-MPI builds when communication between threads
+  is used to mimic MPI calls.
+
+  @author Simon Heybrock
+  @date 2017
+
+  Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_PARALLEL_DLL ThreadingBackend {
+public:
+  ThreadingBackend() = default;
+  explicit ThreadingBackend(const int size);
+
+  ThreadingBackend(const ThreadingBackend &) = delete;
+  const ThreadingBackend &operator=(const ThreadingBackend &) = delete;
+
+  int size() const;
+
+  template <typename... T>
+  void send(int source, int dest, int tag, T &&... args);
+
+  template <typename... T>
+  void recv(int dest, int source, int tag, T &&... args);
+
+  template <typename... T>
+  Request isend(int source, int dest, int tag, T &&... args);
+
+  template <typename... T>
+  Request irecv(int dest, int source, int tag, T &&... args);
+
+private:
+  int m_size{1};
+  std::map<std::tuple<int, int, int>,
+           std::vector<std::unique_ptr<std::stringbuf>>> m_buffer;
+  std::mutex m_mutex;
+};
+
+template <typename... T>
+void ThreadingBackend::send(int source, int dest, int tag, T &&... args) {
+  auto buf = Kernel::make_unique<std::stringbuf>();
+  std::ostream os(buf.get());
+  {
+    // binary_oarchive must be scoped, otherwise the destructor segaults. Since
+    // the buffer is in a unique_ptr that is being moved, maybe this is due to
+    // access from another thread once the lock goes out if scope?
+    boost::archive::binary_oarchive oa(os);
+    oa.operator<<(std::forward<T>(args)...);
+  }
+  std::lock_guard<std::mutex> lock(m_mutex);
+  m_buffer[std::make_tuple(source, dest, tag)].push_back(std::move(buf));
+}
+
+template <typename... T>
+void ThreadingBackend::recv(int dest, int source, int tag, T &&... args) {
+  const auto key = std::make_tuple(source, dest, tag);
+  while (true) {
```

Comment:
Yes, that is what MPI does. I think most implementations just hang indefinitely. Maybe it is possible to configure a timeout, but I am not sure.